After lot of restart due to oom nats panic: corrupt state file

## Defect
On nats cluster affected by 
[https://github.com/nats-io/nats-server/issues/3517](https://github.com/nats-io/nats-server/issues/3517)
afer 1115 restarts on container nats server panic on start.


#### Versions of `nats-server` and affected client libraries used:
[56] 2022/10/10 11:11:13.961798 [INF] Starting nats-server
[56] 2022/10/10 11:11:13.961853 [INF]   Version:  2.9.3-beta.1
[56] 2022/10/10 11:11:13.961859 [INF]   Git:      [cb086bce]
[56] 2022/10/10 11:11:13.961863 [DBG]   Go build: go1.19.1
#### OS/Container environment:
AWS EKS 
#### Steps or code to reproduce the issue:
start 3x nats cluster, with slow disks, create jetstream with file storage and 3 replicas. Push more data than disk are able consume, until nats write disk cache cause container to OOM. Since that nats cluster is in OOM loop after while data on disk are corrupted and nats panic.
#### Expected result:
nats and try to recover from last state
#### Actual result:

nats fail to start:
[142] 2022/10/10 11:00:45.333410 [DBG] Exiting consumer monitor for '$G > STX_SERVER_DATA > DB_SERVICE' [C-R3F-OuTvxCQ0]
panic: corrupt state file

goroutine 51 [running]:
github.com/nats-io/nats-server/v2/server.(*jetStream).applyConsumerEntries(0xc00024c000, 0xc000262d80, 0x0?, 0x0)
	/home/runner/work/nats-server/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:3992 +0x816
github.com/nats-io/nats-server/v2/server.(*jetStream).monitorConsumer(0xc00024c000, 0xc000262d80, 0xc000251680)
	/home/runner/work/nats-server/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:3871 +0xdc6
github.com/nats-io/nats-server/v2/server.(*jetStream).processClusterCreateConsumer.func1()
	/home/runner/work/nats-server/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:3585 +0x25
created by github.com/nats-io/nats-server/v2/server.(*Server).startGoRoutine
	/home/runner/work/nats-server/src/github.com/nats-io/nats-server/server/server.go:3077 +0x85

Noticing this panic as well within K8S using the Helm charts (https://github.com/nats-io/k8s).

No excessive restarts or even OOM. Running three instances, one with around 1 GB and two with around 700 MB RAM used.
Just doing a `kubectl rollout restart nats` to normally restart the instances.

Weirdly enough this happened to two different instances. First nats-2 complained about one consumer, and then nats-1 complained about another one (see below stacktrace).

```
[67] 2022/10/21 09:09:01.145997 [ERR] JetStream cluster could not decode consumer snapshot for 'stream > diff > area-monitor_0' [C-R3F-lzn0afI4]
panic: corrupt state file

goroutine 1578 [running]:
github.com/nats-io/nats-server/v2/server.(*jetStream).applyConsumerEntries(0xc000270000, 0xc00036c480, 0xc002fbda00?, 0x0)
	/home/travis/gopath/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:3992 +0x816
github.com/nats-io/nats-server/v2/server.(*jetStream).monitorConsumer(0xc000270000, 0xc00036c480, 0xc014864b40)
	/home/travis/gopath/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:3871 +0xdc6
github.com/nats-io/nats-server/v2/server.(*jetStream).processClusterCreateConsumer.func1()
	/home/travis/gopath/src/github.com/nats-io/nats-server/server/jetstream_cluster.go:3585 +0x25
created by github.com/nats-io/nats-server/v2/server.(*Server).startGoRoutine
	/home/travis/gopath/src/github.com/nats-io/nats-server/server/server.go:3077 +0x85
```

--- a/server/filestore.go
+++ b/server/filestore.go
@@ -6535,6 +6535,7 @@ func (o *consumerFileStore) stateWithCopy(doCopy bool) (*ConsumerState, error) {
 	return state, nil
 }
 
+// Decode consumer state.
 func decodeConsumerState(buf []byte) (*ConsumerState, error) {
 	version, err := checkConsumerHeader(buf)
 	if err != nil {
@@ -6602,7 +6603,8 @@ func decodeConsumerState(buf []byte) (*ConsumerState, error) {
 				dseq = readSeq()
 			}
 			ts := readTimeStamp()
-			if ts == -1 {
+			// Check the state machine for corruption, not the value which could be -1.
+			if bi == -1 {
 				return nil, errCorruptState
 			}
 			// Adjust seq back.

--- a/server/norace_test.go
+++ b/server/norace_test.go
@@ -5879,3 +5879,23 @@ func TestJetStreamLargeNumConsumersPerfImpact(t *testing.T) {
 	fmt.Printf("\n1000 consumers time is %v\n", tt)
 	fmt.Printf("%.0f msgs/sec\n", float64(toSend)/tt.Seconds())
 }
+
+// Bug when we encode a timestamp that upon decode causes an error which causes server to panic.
+// This can happen on consumer redelivery since they adjusted timstamps can be in the future, and result
+// in a negative encoding. If that encoding was exactly -1 seconds, would cause decodeConsumerState to fail
+// and the server to panic.
+func TestNoRaceEncodeConsumerStateBug(t *testing.T) {
+	for i := 0; i < 200_000; i++ {
+		// Pretend we redelivered and updated the timestamp to reflect the new start time for expiration.
+		// The bug will trip when time.Now() rounded to seconds in encode is 1 second below the truncated version
+		// of pending.
+		pending := Pending{Sequence: 1, Timestamp: time.Now().Add(time.Second).UnixNano()}
+		state := ConsumerState{
+			Delivered: SequencePair{Consumer: 1, Stream: 1},
+			Pending:   map[uint64]*Pending{1: &pending},
+		}
+		buf := encodeConsumerState(&state)
+		_, err := decodeConsumerState(buf)
+		require_NoError(t, err)
+	}
+}

--- a/server/raft.go
+++ b/server/raft.go
@@ -951,6 +951,15 @@ func (n *raft) InstallSnapshot(data []byte) error {
 		return errCatchupsRunning
 	}
 
+	// We don't store snapshots for memory based WALs.
+	// This matches loadSnapshot logic now.
+	// But we should compact.
+	if _, ok := n.wal.(*memStore); ok {
+		n.wal.Compact(n.applied)
+		n.Unlock()
+		return nil
+	}
+
 	var state StreamState
 	n.wal.FastState(&state)
 
@@ -1081,8 +1090,10 @@ func (n *raft) setupLastSnapshot() {
 	n.snapfile = latest
 	snap, err := n.loadLastSnapshot()
 	if err != nil {
-		os.Remove(n.snapfile)
-		n.snapfile = _EMPTY_
+		if n.snapfile != _EMPTY_ {
+			os.Remove(n.snapfile)
+			n.snapfile = _EMPTY_
+		}
 	} else {
 		n.pindex = snap.lastIndex
 		n.pterm = snap.lastTerm

--- a/server/store.go
+++ b/server/store.go
@@ -205,6 +205,7 @@ type ConsumerState struct {
 	Redelivered map[uint64]uint64 `json:"redelivered,omitempty"`
 }
 
+// Encode consumer state.
 func encodeConsumerState(state *ConsumerState) []byte {
 	var hdr [seqsHdrSize]byte
 	var buf []byte
@@ -238,7 +239,7 @@ func encodeConsumerState(state *ConsumerState) []byte {
 
 	// These are optional, but always write len. This is to avoid a truncate inline.
 	if len(state.Pending) > 0 {
-		// To save space we will use now rounded to seconds to be base timestamp.
+		// To save space we will use now rounded to seconds to be our base timestamp.
 		mints := time.Now().Round(time.Second).Unix()
 		// Write minimum timestamp we found from above.
 		n += binary.PutVarint(buf[n:], mints)
@@ -248,7 +249,7 @@ func encodeConsumerState(state *ConsumerState) []byte {
 			n += binary.PutUvarint(buf[n:], v.Sequence-adflr)
 			// Downsample to seconds to save on space.
 			// Subsecond resolution not needed for recovery etc.
-			ts := v.Timestamp / 1_000_000_000
+			ts := v.Timestamp / int64(time.Second)
 			n += binary.PutVarint(buf[n:], mints-ts)
 		}
 	}

